### PR TITLE
Add LoRA support to all linear layers

### PR DESCRIPTION
**Pull request by ertkonuk**
# What does this PR do ?

Add LoRA support to all linear transformations in Attention and MLP modules in NeMo and MCore GPT models.

**Collection**: [Note which collection this PR will affect]
nlp

# Usage
To add LoRa to specific layer, just add the **target_modules** parameter to lora_tuning. The available configuration are below:

```yaml    
lora_tuning:
      target_modules: ['attention_qkv'] # default, adds LoRA to QKV projection layer
      target_modules: ['attention_qkv','attention_dense'] # adds LoRA to QKV projection and Dense layers attention
      target_modules: ['mlp_fc1','mlp_fc2']# adds LoRA to linear layers in the MLP module
      target_modules: ['attention_qkv','attention_dense','mlp_fc1','mlp_fc2'] # adds LoRA to every linear layer in attention and MLP modules
```
There are also shortcut options defined as below:
```yaml    
lora_tuning:
      target_modules: ['attention'] # same as ['attention_qkv','attention_dense']
      target_modules: ['mlp'] # same as  ['mlp_fc1','mlp_fc2']
      target_modules: ['attention','mlp'] # same as ['attention_qkv','attention_dense','mlp_fc1','mlp_fc2']
      target_modules: ['all'] # same as ['attention_qkv','attention_dense','mlp_fc1','mlp_fc2']
```
If the target_modules parameter is not provided by the user, the default value is  
```yaml 
 target_modules: ['attention_qkv']
```
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?
@arendu

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by arendu — DISMISSED** [review]
Looks Good! @ertkonuk have you run some tests offline?

**Review by cuichenx — COMMENTED** [review]
All changes look good to me as well! As long as all code paths are tested offline.

**Comment by arendu** [comment]
jenkins

**Comment by arendu** [comment]
jenkins

**Review by michal2409 — DISMISSED** [review]

**Review by cuichenx — CHANGES_REQUESTED** [review]
unfortunately the forward function of `MCoreSelfAttentionMixin` has become out of date with the current mcore attention forward function... could you update this please?

File: nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py [hunk 1]
```diff
@@ -136,17 +142,85 @@ def get_query_key_value_tensors(self, hidden_states, key_value_states=None):
...
+        output, bias = self.linear_proj(core_attn_out)
+        # LoRA logic
+        if self.is_adapter_available():
+            lora_linear_proj_adapter = self.get_adapter_module(AdapterName.LORA_DENSE_ATTENTION_ADAPTER)
+            if lora_linear_proj_adapter:
+                lora_output = lora_linear_proj_adapter(core_attn_out)
+                output = output + lora_output
+
+        return output, bias
+
```
[the remark this answers is not in the document]
Comment (ertkonuk) [reply nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py]: Thanks for the feedback. I updated MCoreSelfAttentionMixin and MCoreMLPMixin.

**Comment by arendu** [comment]
jenkins

**Review by cuichenx — DISMISSED** [review]
LGTM, thanks!

**Comment by arendu** [comment]
jenkins

**Comment by arendu** [comment]
jenkins

**Comment by arendu** [comment]
jenkins

**Comment by HeyyyyyyG** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Comment by HeyyyyyyG** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Comment by HeyyyyyyG** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Review by cuichenx — APPROVED** [review]

**Comment by cuichenx** [comment]
jenkins

**Comment by arendu** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Review by HeyyyyyyG — APPROVED** [review]

**Comment by HeyyyyyyG** [comment]
jenkins

**Pull request closed without merging**